src/meta/api/src/schema_api_impl.rs

```diff
@@ -1071,6 +1082,222 @@ impl<KV: kvapi::KVApi<Error = MetaError>> SchemaApi for KV {
+
+            let (_, old_virtual_column_opt): (_, Option<VirtualColumnMeta>) =
+                get_pb_value(self, &req.name_ident).await?;
+
+            if old_virtual_column_opt.is_none() {
+                return Err(KVAppError::AppError(AppError::VirtualColumnNotFound(
+                    VirtualColumnNotFound::new(
+                        req.name_ident.table_id,
+                        format!(
+                            "update virtual column with tenant: {} table_id: {}",
+                            req.name_ident.tenant, req.name_ident.table_id
+                        ),
+                    ),
+                )));
+            }
+            let old_virtual_column = old_virtual_column_opt.unwrap();
+
+            let virtual_column_meta = VirtualColumnMeta {
+                table_id: req.name_ident.table_id,
+                virtual_columns: req.virtual_columns.clone(),
+                created_on: old_virtual_column.created_on,
+                updated_on: Some(Utc::now()),
+            };
+
+            // Update virtual column by inserting this record:
+            // (tenant, table_id) -> virtual_column_meta
+            {
+                let if_then = vec![
+                    txn_op_put(&req.name_ident, serialize_struct(&virtual_column_meta)?), /* (tenant, table_id) -> virtual_column_meta */
+                ];
+
+                let txn_req = TxnRequest {
+                    condition: vec![],
+                    if_then,
+                    else_then: vec![],
+                };
+
+                let (succ, _responses) = send_txn(self, txn_req).await?;
+
+                debug!(
+                    req.name_ident = debug(&virtual_column_meta),
+                    succ = display(succ),
+                    "update_virtual_column"
+                );
+
+                if succ {
+                    break;
+                }
+            }
+        }
+
+        Ok(UpdateVirtualColumnReply {})
+    }
+
+    async fn drop_virtual_column(
+        &self,
+        req: DropVirtualColumnReq,
+    ) -> Result<DropVirtualColumnReply, KVAppError> {
+        debug!(req = debug(&req), "SchemaApi: {}", func_name!());
+
+        let ctx = &func_name!();
+        let mut trials = txn_trials(None, ctx);
+        loop {
+            trials.next().unwrap()?;
+
+            let (_, old_virtual_column_opt): (_, Option<VirtualColumnMeta>) =
+                get_pb_value(self, &req.name_ident).await?;
+
+            if old_virtual_column_opt.is_none() {
+                return Err(KVAppError::AppError(AppError::VirtualColumnNotFound(
+                    VirtualColumnNotFound::new(
+                        req.name_ident.table_id,
+                        format!(
+                            "drop virtual column with tenant: {} table_id: {}",
+                            req.name_ident.tenant, req.name_ident.table_id
+                        ),
+                    ),
+                )));
+            }
```

Comment:
Such a routine can be extracted to a function, such as:
https://github.com/datafuselabs/databend/blob/529d184808eecbf1cd09ce31f3629553478f8ef2/src/meta/api/src/util.rs#L417-L437